### ISSUE TITLE
Auth failure when connecting to Azure PostgreSQL

### ISSUE THREAD

**Issue by hassanselim0** [pgbouncer/pgbouncer#325]
When a client tries to connect to PgBouncer and the DB Username has an `@`  character (for example `foo@bar`) then everything after the `@` gets stripped (it becomes just `foo`).
I faced this issue when trying to configure PgBouncer to connect to Azure's managed Postgres service, for some reason it forces usernames to be in the format `user@host`.

Interestingly, if the `[databases]` config has a host and a user specified, the problem still exits. It only goes away if you specify both user and password.

**Comment by davidfetter** [comment]
If you can, please make a small case that reproduces this problem on Azure and on community PostgreSQL. If it only breaks with Azure, this is a matter for Microsoft's technical support.

**Comment by hassanselim0** [comment]
This is really weird, I created a local PostgreSQL instance (via docker) with the exact same credentials of my Azure PostgreSQL instance, and pgbouncer was able to connect to it normally, just changing the `host` value to that of my Azure PostgreSQL instance causes pgbouncer to fail authorization (and adding a `password` value makes it succeed as explained earlier).

Ok so I'll see how I could pass this off to Microsoft's support (or if they have a GitHub repo for issues), but in an effort to make this easier and clearer, I want to know how does the behavior of pgbouncer change between having the `password` field in the `[databases]` section define and not having that field.
After looking at the source code here for an hour (my C/C++ experience is laughable) I think I saw that in case where there is no `password` defined in the db config it fetches it from the `auth_file`, is that correct? If yes, why doesn't it take the password passed in by the client-side connection instead? and how does it read the password from the `auth_file` and send it over to the server? (I couldn't understand that part in the code)

I think the crux of the Azure issue is sending the password, the issue doesn't seem to be with the DB_USER, it just seemed to me (and @ptman) that way because the logs on Azure only show the part before the `@` sign. Maybe it has to do with the `auth_type` being `md5` in my config and Azure doesn't use `md5` so the password doesn't get passed correctly.

**Comment by hassanselim0** [comment]
I've done a few more trials regarding authorization, I tried the following:
```ini
[databases]
* = host=my-pg-db.postgres.database.azure.com port=5432

[pgbouncer]
listen_addr = 0.0.0.0
listen_port = 5432
auth_file = /etc/pgbouncer/userlist.txt
auth_type = plain
```
and the contents of `userlist.txt` is something like `"my-user@my-pg-db" "my-plaintext-password"`.
I did this to test my theory about the issue being in `md5` (previously the auth file had the password as md5), but the authentication still failed. However when I changed the `auth_type` to `plain` it worked!

According to the docs, in case of `auth_type = md5` then the `auth_file` can contain passwords either in plaintext or md5-encrypted, which is why I expected it to work by passing off the password to the server in plaintext if I wrote it that way, but it seems to encrypted first before sending it.

Ok I have a question: The docs say that `auth_type = plain` is deprecated, why is that so? Shouldn't it be fine to send plaintext password if I'm using SSL?
I guess I could use `auth_type = trust`, but I would be better to have the connection to pgbouncer be authenticated (specially for `admin_users`).

So TLDR for anybody who stumbles into this, to make pgbouncer work with Azure you have one of two solutions:
1. Set both `user` and `password` in the `[databases]` section of the config
2. Put the user and password in the `auth_file` in plaintext and set `auth_type = plain`, you can also set it to `trust` but you'll lose admin authentication (very bad thing if publicly exposed)

**Comment by hassanselim0** [comment]
I don't understand the logic behind not wanting to passwords to be stored in plaintext on the server side. Without pgbouncer you would be putting the plaintext password on your servers config files or environment variables. If you don't like that then use a secrets manager and have it inject the passwords in some way or another, which you can probably do with pgbouncer too.

But anyways, it seems like in the case of Azure PostgreSQL, plaintext auth is the only option. So I hope you don't remove that option in the future.

**Comment by HeyHugo** [comment]
Hi, I bumped into this today as well when trying to set up pgbouncer towards a managed azure postgres database. I put a user + password with correct format in the auth_file 

The username has this format: `username@servername`

And like @hassanselim0 said anything after the `@` seems to be stripped so the username is not valid and pgbouncer can't connect:
`WARNING server login failed: FATAL password authentication failed for user "username"`

However I tried setting the user and password under the [databases] section as well as in the userfile and then it works, thanks @hassanselim0!

**Comment by petere** [comment]
These reports are lacking detail. User names containing `@` characters are supported in both PgBouncer and PostgreSQL, and they do appear to work fine.

Please provide:
- exact PostgreSQL setup, starting from `initdb`
- which PostgreSQL roles were created and how
- `pgbouncer.ini` file
- `userlist.txt` file
- `psql` command line for connecting
- PgBouncer and PostgreSQL version (please try with stock PostgreSQL, I can't debug Azure)

**Comment by tetlika** [comment]
having same issue in azure with pgbouncer.x86_64 1.10.0-1.rhel7

md5 does not work, plain works ok

**Comment by petere** [comment]
PostgreSQL on Azure does specific things with this "@servername" business in their user names. I don't know the details of this. This will need someone to dig in and analyze the authentication protocol phase.

**Comment by tetlika** [comment]
@petere can that be prioritised please?

**Comment by petere** [comment]
> @petere can that be prioritised please?

I don't see how. The requested details are not being provided, and AFAICT, the issue is specific to Azure's version of PostgreSQL which is closed source.

**Comment by PJMODOS** [comment]
I wonder if they calculate the password hash differently on azure (omitting the `@severname` part) since the info here suggests that they are actually stripping that part somewhere and the real username is only the part before `@`.

**Comment by petere** [comment]
> I wonder if they calculate the password hash differently on azure (omitting the `@severname` part) since the info here suggests that they are actually stripping that part somewhere and the real username is only the part before `@`.

Hmm, does it work to connect with plain libpq (psql) and md5 authentication? It seems like what you are describing would break libpq.

**Comment by wheezymustafa** [comment]
Dont know if this helps you all, but it looks like Microsoft provides their own image to use pgbouncer with their Postgres offering.  Looks like the md5 encryption isn't supported, which could be why it's breaking here.

**Comment by sak96** [comment]
@tetlika 
My guess is that Azure PostgreSQL is using user map to force the "@ServerName" in username. But pgbouncer does some pre-processing of username and password for md5 auth_type. I will try to recreate this issue in local and update further.

**Comment by tetlika** [comment]
thanks @sak96 , in the end we end up with plain text authentication holding usernames and passwords in plain text in userlist.txt

by the way now we're considering to increase availability in our setup by utilising  read replicas in azure postgres, so scenario looks like this:

we have server1 which currently is master, and username is user@server1

server2 is read  only slave and user for database is user@server2

we add user@server1 and user@server2 into userlist.txt, and when server1 becomes unavailable we can successfully using user@server2 get our data through pgbouncer

however in order to do that we'd need to reconfigure application to use user@server2

do you know if thats possible to somehow overwrite username so in case server1 is unavailable username user@server2 will be used?

thanks

**Comment by petere** [comment]
You can put the user name in the connection string in the `[databases]` section.

**Comment by tetlika** [comment]
thanks for reply @petere so you think this would work? :

* = host=azure1.postgres.database.azure.com port=5432 user=myuser@azure1
* = host=azure2.postgres.database.azure.com port=5432 user=myuser@azure2

and in case myuser@azure1 works ok it will be used, and in case if myuser@azure1 will not work it will automagically failover to myuser@azure2 ?


the thing is reconfiguring pgbouncer or application is not an option for us, the option is to automatically fallback to myuser@azure2 in cause myuser@azure1 fails

**Comment by petere** [comment]
No, that won't work.

It seems that what we're seeing here is that on Azure you can't have the same user name on different instances. That will be pretty problematic for any kind of transparent failover setup.

**Comment by petere** [comment]
I'm closing this issue because the original issue appears to have been resolved: md5 authentication is not supported by PostgreSQL on Azure.

**Comment by thesuperzapper** [comment]
For those watching this, you might want to consider setting `auth_type` to `scram-sha-256` rather than `trust`, otherwise with `auth_type = trust`, anyone can use your PgBouncer to connect to your PostgreSQL instance (even if they don't provide a password, as it uses your `users.txt`).

I have tested `auth_type = scram-sha-256` with Azure PostgreSQL, and it seems to work fine. 
(Just be careful to ensure your clients support  `scram-sha-256`)

---

I also note that `auth_type = plain` works, but results in sending your passwords un-hashed over the wire.

**Comment by thesuperzapper** [comment]
Also, @petere is correct, Azure PostgreSQL does not support `md5` for password authentication, which is what causes this issue.

**Comment by AmandinePingetPhD** [comment]
I had the same issue, I solved it by removing "@hostname" and keeping only "user" with md5 authentification. Hope it helps!

**Comment by jerynmathew** [comment]
@AmandinePingetPhD Can you give an exact example of how you solved this? Do you change in the ini file and the userlist.txt? can u share a pseudo sample for my understanding.

**Comment by AmandinePingetPhD** [comment]
@jerynmathew  I don't remember precisely my exact configuration and I can't share it but as far as I remember, I made changes in the pg_ident.conf or pg_hba.conf specifying user without "@hostname" with md5 authentication. Hope it can help (but sorry, it's far away for me :s)
